parlai/core/build_data.py

```diff
@@ -24,6 +24,26 @@
 from multiprocessing import Pool
 
 
+def download_check(dpath, urls, file_names, sha256, from_google=False):
```

Comment:
explicitly put checksum in the name please.

can we maybe define a struct like

```
class DownloadableFile:
     url: String = ""
     sha256: String = ""
```

or something?

Reply:
I'd like to be imposing *more* structure and rigor in this part of the code base, not just adding another method.

Reply:
Just to make sure, do you want each task to have an object of this class and the object be the input for the above download and checksum method?

Reply:
Yeah I think we should be targeting a class that corresponds to a URL and a checksum. Then each teacher can have a fixed list of these objects.

Ideally then the download stuff would be a method in the class, but even just having the well defined data structure and that was the one thing you passed would be way better than where parlai is today.

Your solution is strong, but doesn’t provide us with additional structure that we can leverage to build tests.

Reply:
I want to have some confirmation before going ahead and changing all the files, so is a class like this something you would want to be implemented? Let me know if there is anything missing / should be included in the class.
```
class DownloadableFile:
    def __init__(self, url, file_name, hashcode, zipped=True, from_google=False):
        self.url = url
        self.file_name = file_name
        self.hashcode = hashcode
        self.zipped = zipped
        self.from_google = from_google
    
    def checksum(self, dpath):
        sha256_hash = hashlib.sha256()
        with open(os.path.join(dpath, self.file_name), "rb") as f:
            for byte_block in iter(lambda: f.read(65536), b""):
                sha256_hash.update(byte_block)
            if sha256_hash.hexdigest() != self.hashcode:
                # remove_dir(dpath)
                raise ValueError("Checksum Failed")
            else:
                print("[ Checksum Successful ]")

    def download_file(self, dpath):
        if self.from_google:
            download_from_google_drive(self.url, os.path.join(dpath, self.file_name))
        else:
            download(self.url, dpath, self.file_name)
        
        self.checksum(dpath)

        if self.zipped:
            untar(dpath, self.file_name)
```

With the given class, each task will have a FILE_LIST that we an iterate through to check all the URLS.

```
URLS = ['https://github.com/deepmind/AQuA/archive/master.zip']
FILE_NAMES = ['aqua.zip']
SHA256 = ['08ea725477f6a8577a7cc1a2ae08c7a56917aa3ec45193f173b298b6b526c603']

FILE_LIST = [DownloadableFile(url, file_name, hashcode) for url, file_name, hashcode in zip(URLS, FILE_NAMES, SHA256)]

def build(opt):
    dpath = os.path.join(opt['datapath'], 'AQuA')
    version = '1.0'

    if not build_data.built(dpath, version_string=version):
        print('[building data: ' + dpath + ']')

        if build_data.built(dpath):
            # an older version exists, so remove these outdated files.
            build_data.remove_dir(dpath)
        build_data.make_dir(dpath)

        for f in file_list:
            f.download_file(dpath)

        base_path = os.path.join(dpath, 'AQuA-master')
        new_path = os.path.join(dpath, 'AQuA')

        shutil.move(base_path, new_path)

        # mark the data as built
        build_data.mark_done(dpath, version_string=version)
```

Reply:
Yes I like that. The only thing i'd suggest is instead have the items explicitly constructed, rather than zipping through the list:

```python
RESOURCES = [
    DownloadableFile(
         'https://github.com/deepmind/AQuA/archive/master.zip',
         'aqua.zip',
         '08ea725477f6a8577a7cc1a2ae08c7a56917aa3ec45193f173b298b6b526c603'
    ),
]
```

the nice thing is that this also enables us to implement

```python 
class DownloadableFile:
    # ...
    def check_header(self):
        # do the equivalent of curl -I, where you ask if the resource exists,
        # without downloading it
```

and that would enable us to also sort out #1875.